src/gf.c

```diff
@@ -1252,8 +1241,11 @@ jl_methlist_t *jl_method_list_insert(jl_methlist_t **pml, jl_tupletype_t *type,
             l->va = jl_is_va_tuple(type);
             l->isstaged = isstaged;
             l->invokes = (struct _jl_methtable_t *)jl_nothing;
+            if (l->func)                  // TODO jb/functions
+                jl_gc_preserve((jl_value_t*)l->func);  // keep roots for replaced method
```

Comment:
i think you know this because of the comment, but `jl_gc_preserve` may not actually do very well at preserving the value